session/session.go

```diff
@@ -96,14 +96,25 @@ func newZKSession(servers string, recvTimeout time.Duration, logger stdLogger, c
 	}
 
 	var event zookeeper.Event
-	select {
-	case event = <-events:
-		if event.State != zookeeper.STATE_CONNECTED {
-			return nil, ErrZKSessionNotConnected
+
+	Loop:
+		for {
```

Comment:
I think this would read a lot better if you extract the logic into a separate function. Something like:
```golang
func waitForConnection(events <-chan zookeeper.Event) error {
  for {
    select {
    case event := <-events:
      switch event.State {
        case zookeeper.STATE_AUTH_FAILED:
        case zookeeper.STATE_EXPIRED_SESSION:
        case zookeeper.STATE_CLOSED:
          return ErrZKSessionNotConnected
        case zookeeper.STATE_CONNECTED:
          return nil
      }
    case <-time.After(5 * time.Second):
      return ErrZKSessionNotConnected
    }
  }
  // We should not get to this state
  return ErrZKSessionNotConnected
}
```